### PR TITLE
feat: upgrade prometheus

### PR DESCRIPTION
**Pull request by pracucci**
**What this PR does / why we need it**:

Prometheus introduced a breaking change to storage interface, adding `LabelsHint` to `LabelNames()` and `LabelValues()`. This is breaking prometheus vendoring in GEM repository, where we also have an application (gateway) vendoring Loki.

In this PR I've attempted a prometheus upgrade in Loki. Details:

- It's carrying a bunch of other dependencies upgraded too 😱 
- gRPC upgrade comes with the deprecation of `grpc.Dial()`. There's no trivial replacement, so I've added a nolint rule for now (we've done the same in Mimir to gain some time).
- Recent changes in Prometheus regexp label matcher makes it incomparable with a deep equal. There are few tests in Loki running a deep equal on object containing regexp matchers. I've changed these tests to compare the matchers using the `String()` representation (same approach took in Mimir). Don't know if this is acceptable to you.

Note to reviewers:
- I need help fixing `clients/pkg/logentry/logql/TestParse`. I "fixed" it comparing only matchers, but it lowers the confidence of the test. I only came up with complex ideas to solve it, e.g. re-implement a deep equal using a visitor pattern, or use reflection to manipulate the regexp matcher under the hood before running a deep equal.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/logql/engine.go [hunk 1]
```diff
@@ -444,11 +443,6 @@ func (q *query) JoinSampleVector(next bool, r StepResult, stepEvaluator StepEval
 		return vec, nil
 	}
 
-	stepCount := int(math.Ceil(float64(q.params.End().Sub(q.params.Start()).Nanoseconds()) / float64(q.params.Step().Nanoseconds())))
```
Comment (pracucci) [inline pkg/logql/engine.go:447]: Note to reviewers: the linter complained this in ineffectual assignment and I think it's right.

File: clients/pkg/promtail/targets/docker/targetmanager_test.go [hunk 2]
```diff
@@ -50,7 +50,7 @@ func Test_TargetManager(t *testing.T) {
 		case strings.HasSuffix(path, "/networks"):
 			// Serve networks
 			w.Header().Set("Content-Type", "application/json")
-			err := json.NewEncoder(w).Encode([]types.NetworkResource{})
```
Comment (pracucci) [inline clients/pkg/promtail/targets/docker/targetmanager_test.go:53]: Note to reviewers: `types.NetworkResource` is deprecated, but `network.Inspect` is expected to be the same exact type.

File: pkg/logql/engine.go (hunk 1 above)
Comment (vlad-diachenko) [reply pkg/logql/engine.go:447]: I also do not see any usage of this variable

**Review by vlad-diachenko — APPROVED** [review]
Looks great. Thanks @pracucci

**Comment by pracucci** [comment]
> Looks great. Thanks @pracucci

Well, you did the hard work, so thanks @vlad-diachenko !

**Pull request merged**